### PR TITLE
Support for LDAPS | FATE#316086

### PR DESCRIPTION
**Pull request by vmoravec**
It gives the user to setup the following option in `/etc/openldap/ldap.conf`:
`URI  ldaps://secure.login.cz`

### CODE REVIEW

File: src/include/ldap/ui.rb [hunk 1]
```diff
@@ -210,86 +210,221 @@ def check_certificate(file)
       true
     end
 
+    def switch_ssl_config_widgets mode
+      switch =
+        case mode
+          when :on  then true
+          when :off then false
+        end
```
Comment (jsuchome) [inline src/include/ldap/ui.rb:218]: Isn't there ternary operator in ruby?

File: src/include/ldap/ui.rb [hunk 2]
```diff
@@ -210,86 +210,221 @@ def check_certificate(file)
...
+                                      _("StartTLS (default port 389)"),
+                                      use_tls
+                                    )
+                                  ),
+                                  Left(
+                                    RadioButton(
+                                      Id(:use_ldaps),
+                                      Opt(:notify),
+                                      _("LDAPS (default port 636)"),
+                                      use_ldaps
```
Comment (jsuchome) [inline src/include/ldap/ui.rb:287]: Do you need to list port numbers in the label? I find it a bit confusing...

File: src/include/ldap/ui.rb [hunk 3]
```diff
@@ -210,86 +210,221 @@ def check_certificate(file)
...
           ),
-          HSpacing(1)
+          VSpacing(0.4)
+        ),
+        HSpacing(1)
         )
       )
-      ret = :again
+
+      switch_ssl_config_widgets(:off) unless Ldap.use_secure_connection?
```
Comment (jsuchome) [inline src/include/ldap/ui.rb:386]: what about using Ldap.use_secure_connection as argument for switch_ssl_config_widgets?

Comment (vmoravec) [inline src/include/ldap/ui.rb:386]: I like being more explicit than just by passing a param as boolean.

File: src/include/ldap/ui.rb (hunk 2 above)
Comment (vmoravec) [inline src/include/ldap/ui.rb:287]: Ok, should this information go into some help or should I remove it completely? I'm not sure what's the best practice for yast UI dialogs.

Comment (jsuchome) [inline src/include/ldap/ui.rb:287]: Help is a good place, IMHO.

**Comment by vmoravec** [comment]
Help text added, @jsuchome pls have a look.

**Comment by jsuchome** [comment]
LGTM (did not test the code)

File: src/include/ldap/ui.rb (hunk 1 above)
Comment (jreidinger) [inline src/include/ldap/ui.rb:218]: or even better `switch = mode == :on`

File: src/include/ldap/ui.rb [hunk 4]
```diff
@@ -210,86 +210,221 @@ def check_certificate(file)
...
-                  PushButton(Id(:import_cert), _("Do&wnload CA Certificate"))
+            Frame(
+              _("SSL/TLS Configuration"),
+              HBox(
+                VBox(
+                  Left(
+                    CheckBox(
+                      Id(:secure_ldap),
+                      Opt(:notify),
+                      "Use SSL/TLS",
```
Comment (jreidinger) [inline src/include/ldap/ui.rb]: missing translation

**Comment by vmoravec** [comment]
I tested the code with LDAP server available at login-internal.suse.de , which support ldaps only (with `Request server certificate` disabled)

File: src/include/ldap/ui.rb (hunk 3 above)
Comment (jreidinger) [inline src/include/ldap/ui.rb:386]: well, then better option is to have boolean parameter for switch and explicit `enable_ssl_config_widgets` and `disable_ssl_config_widgets` that pass boolean below. It is more readable without additional symbol.

File: src/include/ldap/ui.rb [hunk 5]
```diff
@@ -210,86 +210,221 @@ def check_certificate(file)
...
+
+        when :use_ldaps
+          use_ldaps = true
+          use_tls = false
+          Ldap.modified = true
+          UI.ChangeWidget(Id(:request_server_certificate), :Value, true)
+
+        when :request_server_certificate
+          request_server_certificate =
+            case UI.QueryWidget(Id(:request_server_certificate), :Value)
```
Comment (jreidinger) [inline src/include/ldap/ui.rb]: here is ternary operator more understandable.

```
widget_checked = UI.QueryWidget(Id(:request_server_certificate), :Value)
request_server_certificate = widget_checked ? 'demand' : 'allow'
```

File: src/include/ldap/ui.rb (hunk 3 above)
Comment (vmoravec) [inline src/include/ldap/ui.rb:386]: I don't think the method reads bad, it does what it says: switch the config widgets off or on . I had similar enable/disable wrappers there and merged it into a single method, it's just matter of preference and the code is really simple to misunderstand what it does.

File: src/modules/Ldap.rb [hunk 6]
```diff
@@ -588,9 +605,15 @@ def ShortSummary
...
         summary = Ops.add(
           Ops.add(summary, "<br>"),
-          _("LDAP TLS/SSL Configured")
+          _("LDAP TLS Configured")
         )
       end
+
+      if @ldaps
+        summary << "<br/>"
+        summary + _("LDAPS Configured")
```
Comment (jreidinger) [inline src/modules/Ldap.rb]: why use `<<` above and here `+` which looks useless as it is not stored?

File: src/modules/Ldap.rb [hunk 7]
```diff
@@ -751,11 +774,21 @@ def ReadLdapHosts
...
         ret = ReadLdapConfEntry("host", "")
       else
+        detect_ldaps(uri)
         ret = uri2servers(uri)
       end
       ret
     end
 
+    def detect_ldaps uri
+      @ldaps = !!(uri.match(/\Aldaps/))
```
Comment (jreidinger) [inline src/modules/Ldap.rb]: There is URI library in ruby which can return you scheme to check

File: src/modules/Ldap.rb [hunk 8]
```diff
@@ -751,11 +774,21 @@ def ReadLdapHosts
...
+      @request_server_certificate = read_openldap_config('TLS_REQCERT').first
+    end
+
+    def detect_uri_scheme
```
Comment (jreidinger) [inline src/modules/Ldap.rb:789]: why it is called detect uri scheme when it doesn't detect jsut return protocol part of URI?

Comment (vmoravec) [inline src/modules/Ldap.rb:789]: The protocol part of uri is called scheme, I think.

File: src/modules/Ldap.rb [hunk 9]
```diff
@@ -2216,82 +2249,34 @@ def WriteLDAP(objects)
...
-      if open_uri == []
-        open_uri = Convert.to_list(
-          SCR.Read(path(".etc.ldap_conf.v.\"/etc/openldap/ldap.conf\".URI"))
-        )
-      end
-      if open_uri == []
-        open_host = Convert.to_list(
-          SCR.Read(path(".etc.ldap_conf.v.\"/etc/openldap/ldap.conf\".host"))
-        )
+      uris = @server.split.map {|u| "#{detect_uri_scheme}#{u}" }.join(' ')
```
Comment (jreidinger) [inline src/modules/Ldap.rb]: better to have 
`{|server| uri_scheme + server }`
then
`{|u| "#{detect_uri_scheme}#{u}" }`

File: src/modules/Ldap.rb (hunk 7 above)
Comment (vmoravec) [inline src/modules/Ldap.rb]: Good point

File: src/include/ldap/ui.rb [hunk 10]
```diff
@@ -210,86 +210,218 @@ def check_certificate(file)
...
     end
 
+    def switch_ssl_config_widgets mode
+      switch =
+        case mode
+          when :on  then true
+          when :off then false
+        end
+
+      widgets = [
```
Comment (jreidinger) [inline src/include/ldap/ui.rb]: looks like constant

File: src/modules/Ldap.rb [hunk 11]
```diff
@@ -588,9 +606,14 @@ def ShortSummary
         # summary
         summary = Ops.add(
           Ops.add(summary, "<br>"),
-          _("LDAP TLS/SSL Configured")
+          _("LDAP TLS Configured")
         )
       end
+
+      if @ldaps
+        summary + "<br/>" + _("LDAPS Configured")
```
Comment (jreidinger) [inline src/modules/Ldap.rb]: does it have special side effect or result of sum is vanished?

File: src/modules/Ldap.rb [hunk 12]
```diff
@@ -2868,7 +2854,7 @@ def Write(abort)
         WriteLdapConfEntry("host", nil)
         uri = Builtins.mergestring(
           Builtins.maplist(Builtins.splitstring(@server, " \t")) do |u|
-            Ops.add("ldap://", u)
+            detect_uri_scheme + u
           end,
           " "
```
Comment (jreidinger) [inline src/modules/Ldap.rb:2859]: I see there third time similar method for parsing server and merging it with `" "` or `","`

Comment (vmoravec) [inline src/modules/Ldap.rb:2859]: Can be, there are a lot of places with code duplication, I fixed only what was necessary for the rest to work with my additions.

File: src/modules/Ldap.rb (hunk 11 above)
Comment (vmoravec) [inline src/modules/Ldap.rb]: Is vanished.. no special side effect. Will fix it.

File: src/include/ldap/ui.rb (hunk 10 above)
Comment (vmoravec) [inline src/include/ldap/ui.rb]: Looks so, but it's not, will change the code to iterate over the collection directly without intermediate variable.

**Pull request merged**